crates/wasi-async-runtime/src/streams.rs

```diff
@@ -0,0 +1,234 @@
+use futures_lite::{io, AsyncBufRead, AsyncRead, AsyncWrite};
+use wasi::io::streams::{
+    InputStream as WasiInputStream, OutputStream as WasiOutputStream,
+    StreamError as WasiStreamError,
+};
+
+const DEFAULT_BUF_LEN: usize = 32;
+
+#[derive(Debug)]
+/// Wraps [wasi::io::streams::InputStream] to enable usage with async libraries.
+pub struct InputStream {
+    inner: WasiInputStream,
+    poll_handle: Option<PollHandle>,
+    buf: RefCell<Vec<u8>>,
+}
+
+impl InputStream {
+    /// Instatiate the stream.
+    ///
+    /// # Examples
+    ///
+    /// ```ignore
+    /// block_on(|r| {
+    ///     let lines = InputStream::new(my_wasi_input_stream, r.clone()).lines();
+    ///     while let Some(line) = lines.next().await.transpose()? {
+    ///         eprintln!("Got a line: {line}");
+    ///     }
+    /// });
+    /// ```
+    pub fn new(inner: WasiInputStream, reactor: Reactor) -> Self {
+        let poll_handle = Some(reactor.register(inner.subscribe()));
+        Self {
+            inner,
+            poll_handle,
+            buf: RefCell::default(),
+        }
+    }
+
+    fn poll(&self, cx: &mut task::Context<'_>) -> task::Poll<()> {
+        self.poll_handle.as_ref().unwrap().poll(cx)
+    }
+}
+
+impl Drop for InputStream {
+    fn drop(&mut self) {
+        // NOTE: we can't utilize [Drop] for [PollHandle] given the resource hierarchy
+        self.poll_handle.take().unwrap().unregister();
+    }
+}
+
+impl AsyncRead for InputStream {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut task::Context<'_>,
+        buf: &mut [u8],
+    ) -> task::Poll<std::io::Result<usize>> {
+        if self.poll(cx).is_pending() {
+            return task::Poll::Pending;
+        }
+
+        let mut len = self.buf.borrow().len();
+        let bytes = match len {
+            0 => match self.inner.read(buf.len() as u64) {
+                Ok(bytes) => bytes,
+                Err(WasiStreamError::Closed) => return task::Poll::Ready(Ok(0)),
+                Err(WasiStreamError::LastOperationFailed(err)) => {
+                    return task::Poll::Ready(Err(io::Error::new(
+                        io::ErrorKind::BrokenPipe,
+                        err.to_debug_string(),
+                    )))
+                }
+            },
+            _ => {
+                if buf.len() < len {
+                    len = buf.len();
+                }
+                self.buf.borrow_mut().drain(0..len).collect()
+            }
+        };
+        let len = bytes.len();
+        bytes.into_iter().enumerate().for_each(|(i, b)| buf[i] = b);
+        task::Poll::Ready(Ok(len))
+    }
+}
+
+impl AsyncBufRead for InputStream {
+    fn poll_fill_buf(
+        self: Pin<&mut Self>,
+        cx: &mut task::Context<'_>,
+    ) -> task::Poll<std::io::Result<&[u8]>> {
+        let this = unsafe { &*Pin::<&mut Self>::get_unchecked_mut(self) };
+        let buf = unsafe { this.buf.as_ptr().as_mut().unwrap() };
```

Comment:
While I _think_ these `unsafe` calls don't violate the `Pin` contract, it'd be good to get a second opinion on that.

Reply:
It turns out these `unsafe`s can be offloaded to the `bytes` crate.